src/libfetchers/fetchers.cc

```diff
@@ -52,8 +52,38 @@ Attrs Input::toAttrs() const
     return attrs;
 }
 
+Tree Input::substituteTree(ref<Store> store) const
+{
+    assert(narHash);
+    auto storePath = store->makeFixedOutputPath(true, narHash.value(), "source");
+
+    store->ensurePath(storePath);
+
+    debug("using substituted/cached input '%s' in '%s'",
+        to_string(), store->printStorePath(storePath));
+
+    auto actualPath = store->toRealPath(storePath);
+
+    return {
+        Tree {
+            .actualPath = actualPath,
+            .storePath = std::move(storePath),
+            .info = { narHash.value() }
+        }
+    };
+}
+
 std::pair<Tree, std::shared_ptr<const Input>> Input::fetchTree(ref<Store> store) const
 {
+    if (narHash) {
+        try {
+            auto tree = substituteTree(store);
+            return {std::move(tree), shared_from_this()};
```

Comment:
Unfortunately, this is not correct because this will return a different `Tree` and `Input` than `fetchTreeInternal()`. For instance, for a Git input, `fetchTreeInternal()` will return an `Input` that has `rev` set, and a `Tree` that has `revCount` and `lastModified` set.

So we can only substitute the store path if we can get those `Input` / `Tree` attributes in some other way. It works for flake inputs because `flake.lock` provides the missing attributes.

Reply:
Ok makes sense. So basically a locked ref always needs to be returned? Not that you couldn't substitute an unlocked ref with a narHash, but that wouldn't allow things like lockfile creation? If so was wondering why LockedRef isn't a subclass, which fetchTree could then return.